src/streaming.cairo

```diff
@@ -0,0 +1,219 @@
+        end_time: u64,
+        total_amount: u128
+    );
+
+    fn claim_stream(
+        ref self: TContractState,
+        streamer: ContractAddress,
+        recipient: ContractAddress,
+        start_time: u64,
+        end_time: u64,
+     );
+
+     fn cancel_stream(
+         ref self: TContractState,
+         recipient: ContractAddress,
+         start_time: u64,
+         end_time: u64
+     );
+
+    fn get_stream_info(
+        ref self: TContractState, 
+        streamer: ContractAddress,
+        recipient: ContractAddress,
+        start_time: u64,
+        end_time: u64,
+    ) -> (u128, u128);
+}
+
+#[starknet::component]
+mod streaming {
+    use konoha::contract::Governance;
+    use konoha::contract::{IGovernanceDispatcher, IGovernanceDispatcherTrait};
+    use konoha::traits::{IGovernanceTokenDispatcher, IGovernanceTokenDispatcherTrait};
+    use starknet::ContractAddress;
+    use starknet::{get_block_timestamp, get_caller_address, get_contract_address};
+
+    #[storage]
+    struct Storage{
+        streams: LegacyMap::<(ContractAddress, ContractAddress, u64, u64), (u128, u128)> // (claimed_amount, total_amount)
+    }
+
+    #[derive(starknet::Event, Drop, Serde)]
+    #[event]
+    enum Event{
+        StreamCreated: StreamCreated,
+        StreamClaimed: StreamClaimed,
+        StreamCanceled: StreamCanceled
+    }
+
+    #[derive(starknet::Event, Drop, Serde)]
+    struct StreamCreated{
+        streamer: ContractAddress,
+        recipient: ContractAddress,
+        start_time: u64,
+        end_time: u64,
+        total_amount: u128
+    }
+
+    #[derive(starknet::Event, Drop, Serde)]
+    struct StreamClaimed{
+        streamer: ContractAddress,
+        recipient: ContractAddress,
+        start_time: u64,
+        end_time: u64,
+        total_amount: u128
+    }
+
+    #[derive(starknet::Event, Drop, Serde)]
+    struct StreamCanceled{
+        streamer: ContractAddress,
+        recipient: ContractAddress,
+        start_time: u64,
+        end_time: u64,
+        reclaimed_amount: u256        
+    }
+    
+    //TODO:
+    #[embeddable_as(StreamingImpl)]
+    impl Streaming<
+        TContractState, +HasComponent<TContractState>
+    > of super::IStreaming<ComponentState<TContractState>> {
+        
+        fn add_new_stream(
+            ref self: ComponentState<TContractState>,
+            streamer: ContractAddress,
+            recipient: ContractAddress,
+            start_time: u64,
+            end_time: u64,
+            total_amount: u128
+        ) {
+            let key = (get_caller_address(), recipient, start_time, end_time);
+
+            assert(get_caller_address() == get_contract_address(), 'not self-call');
+            assert(start_time < end_time, 'starts first');
+
+            let mut claimable_amount = 0;
+            self.streams.write(key, (claimable_amount, total_amount));
+        
+            self.emit(StreamCreated{
+                streamer: get_caller_address(),
```

Comment:
I was saving the streamer so I could have a place holder for who was calling. You're right I don't need it as a parameter in adding a new stream if its an emitted event and the caller is the contract. i just removed it in add_new_stream and everything still works. is this what you meant?

Reply:
Do we have to have the streamer in storage if it's always governance?

We will, however, need/want streaming in treasury too – that shuffles the things quite a bit.

Reply:
Fixed issues,

Pushed a version without streamer storage. (also fixes the recipient and streamer swap)

In terms of the treasury, would the streamer also not be the governance? What would the treasury need that the code doesn't have?

Reply:
Yes, the streamer would be treasury (that's where the money would come from), but it would be approved by governance. So tokens wouldn't be minted, but transferred from the streaming contract itself. (can't mint ETH/USDC/STRK).